samza-core/src/main/scala/org/apache/samza/checkpoint/OffsetManager.scala

```diff
@@ -274,6 +293,25 @@ class OffsetManager(
         }
       }
     }
+
+    // delete corresponding startpoints after checkpoint is supposed to be committed
+    if (startpointManager != null && startpoints.contains(taskName)) {
+      val sspStartpoints = checkpoint.getOffsets.keySet.asScala
+        .intersect(startpoints.getOrElse(taskName, Map.empty[SystemStreamPartition, Startpoint]).keySet)
+
+      // delete startpoints for this task and the intersection of SSPs between checkpoint and startpoint.
+      sspStartpoints.foreach(ssp => {
+        startpointManager.deleteStartpointForTask(ssp, taskName)
+        info("Deleted startpoint for SSP: %s and task: %s" format (ssp, taskName))
+      })
+      startpoints -= taskName
+
+      if (startpoints.isEmpty) {
```

Comment:
Should we leave the StartpointManager running so that new startpoints can be written?

Reply:
What if someone wanted to write some more Startpoints before restarting?

Reply:
That makes sense that the job wouldn't apply the new Startpoints until the next restart.
However, does stopping the StartpointManager prevent it from writing the new Startpoints? To me, if something gets stopped, then none of its functionality should be expected to work.

Reply:
It shouldn't matter here. Only reads and deletes happen in the OffsetManager. Once all of that is completed, no other startpoint operation is expected. The writes would originate externally from a tool or service that writes the startpoints into the store.

Reply:
Ah, thanks for clarifying the intention that writes would come from somewhere else.
Would it be simpler though to just let it close when the container shuts down? Is there a benefit in shutting down early?

Reply:
It would be simpler, but the underlying store may keep an open idle connection for no reason. If we can free up resources when not in use, I think that'd be better.